packages/muelu/src/Misc/MueLu_FilteredAFactory_def.hpp

```diff
@@ -708,6 +734,226 @@ namespace MueLu {
+                                           // ratio. Do we want to allow the diagonal
+                                           // to go negative, just to have a better ddom
+                                           // ratio? This current choice essentially
+                                           // changes 'Target' to a large number
+                                           // meaning that we will allow the new
+                                           // ddom number to be fairly large (because
+                                           // the old one was infinity)
+
+        ArrayView<const SC> tvals;
+        A.getLocalRowView(row, inds, vals);           
+        size_t nnz = inds.size();
+        if (nnz == 0) continue;
+        filteredA.getLocalRowView(row, finds, tvals);//assume 2 getLocalRowView()s 
+                                                     // have things in same order
+        fvals = ArrayView<SC>(const_cast<SC*>(tvals.getRawPtr()), nnz);
+
+        LO diagIndex = -1;
+
+        PosOffSum=zero; NegOffSum=zero; PosOffDropSum=zero; NegOffDropSum=zero;
+        diag=zero; NumPosKept=0; NumNegKept=0;
+
+        // first record diagonal, offdiagonal sums and off diag dropped sums
+        for (size_t j = 0; j < nnz; j++) {
+          if (inds[j] == row) {
+            diagIndex = j;
+            diag = vals[j];
+          }
+          else { // offdiagonal
+            if (TST::real(vals[j]) > TST::real(zero) ) {
+              PosOffSum += vals[j];
+              if (TST::magnitude(fvals[j]) == TST::magnitude(zero) ) // not in filtered matrix
+                PosOffDropSum += vals[j];
+              else NumPosKept++;
+            }
+            else {
+              NegOffSum += vals[j];
+              if (TST::magnitude(fvals[j]) == TST::magnitude(zero) ) // not in filtered matrix
+                NegOffDropSum += vals[j];
+              else NumNegKept++;
+            }
+          }
+        }
+
+        // measure of diagonal dominance if no lumping is done. 
+        if (TST::magnitude(diag) != TST::magnitude(zero)  )
+          noLumpDdom = (PosOffSum - NegOffSum)/diag;
+
+        // Target is an acceptable diagonal dominance ratio
+        // which should really be larger than 1
+
+        Target = rho*noLumpDdom;
+        if (TST::magnitude(Target) <= TST::magnitude(rho)) Target = rho2;//rstumin change
+
+        PosFilteredSum = PosOffSum - PosOffDropSum;
+        NegFilteredSum = NegOffSum - NegOffDropSum;
+           // Note: PosNotFilterdSum is not equal to the sum of the
+           // positive entries after lumping. It just reflects the
+           // pos offdiag sum of the filtered matrix before lumping
+           // and does not account for negative dropped terms lumped
+           // to the positive kept terms.
+
+        // dropped positive offdiags always go to the diagonal as these
+        // always improve diagonal dominance.
+
+        diag += PosOffDropSum;// rstumin check in other code? 
```

Comment:
remove this comment

Reply:
@jhux2 removed this comment and added a few more